index.ts

```diff
@@ -258,6 +256,56 @@ export class Project {
       dep.baseDir = path.join(this.baseDir, 'node_modules', dep.name);
       dep.writeSync();
     }
+    for (let [name, { dir: target }] of this.dependencyLinks) {
+      this.writeLinkedPackage(name, target);
+    }
+  }
+
+  private writeLinkedPackage(name: string, target: string) {
+    let targetPkg = require(path.join(target, 'package.json'));
+    let peers = new Set(Object.keys(targetPkg.peerDependencies ?? {}));
+    let destination = path.join(this.baseDir, 'node_modules', name);
+
+    if (peers.size === 0) {
+      // no peerDeps, so we can just symlink the whole package
+      fs.ensureSymlinkSync(target, destination, 'dir');
+      return;
+    }
+
+    // need to reproduce the package structure in our own location
+    this.hardLinkContents(target, destination);
+
+    for (let section of ['dependencies', 'peerDependencies']) {
+      if (targetPkg[section]) {
+        for (let depName of Object.keys(targetPkg[section])) {
+          if (peers.has(depName)) {
+            continue;
+          }
+          let depTarget = resolvePackagePath(depName, target);
+          if (!depTarget) {
+            throw new Error(`package ${name} in ${target} depends on ${depName} but we could not resolve it`);
+          }
+          fs.ensureSymlinkSync(
```

Comment:
Although `fs.ensure*` methods can be handy, i find they can easily paper over actual issues (both correctness and performance). Is it ok if we are more explicit, and explicitly create dir paths before linking into them? Reading this code, it should be totally known when it is correct to do so.

But let me know what you think.

Reply:
I don't really want to change this. The ensure methods make us more robust to unexpected things on the filesystem, and we don't really have sole control over the filesystem.

Reply:
Thats ok if you feel strongly.

In this case, it is a region of the FS that we do have control over. But its not a big deal.